frontend/app/templates/index.html

```diff
@@ -0,0 +1,507 @@
+                    queryParams.jsonb = JSON.stringify(filters);
+                }
+
+                queryRecords(queryParams);
+            }
+
+            function getJsonbFilters() {
+                // Read filters from the form and return them as an object
+                // that we can use to query Ashlar.
+                let filters = {};
+                let formValues = $('#filter-form').serializeArray();
+
+                formValues.forEach((field) => {
+                    // Form data gets encoded as an array of objects storing
+                    // the name and value of the field. Process that object
+                    // into a JSON object suitable for querying.
+                    let name = field.name,
+                        value = field.value;
+
+                    let filterableField = name !== 'record_type';
+
+                    if (value && filterableField) {
+                        // Field names are namespaced like so:
+                        //
+                        // 'schemaName#fieldName#queryRule`
+                        //
+                        // e.g.:
+                        //
+                        // 'incidentDetails#Select List#select'
+                        //
+                        // Parse this structure to format the query.
+                        let fieldParts = name.split('#');
+                        let schemaName = fieldParts[0],
+                            fieldName = fieldParts[1],
+                            queryRule = fieldParts[2];
+
+                        // If this is the first iteration in the loop, there
+                        // won't yet be a key for the schema in the filter object.
+                        if (!filters[schemaName]) {
+                            filters[schemaName] = {};
+                        }
+
+                        // If this is a number query (or another type of field
+                        // that accepts multiple inputs, like a multiselect)
+                        // the nested query object may already exist, so check
+                        // for it before creating it.
+                        let nestedQuery = {};
+                        if (!filters[schemaName][fieldName]) {
+                            filters[schemaName][fieldName] = {};
+                        } else {
+                            nestedQuery = filters[schemaName][fieldName];
+                        }
+
+                        // Construct the actual query.
+                        switch(queryRule) {
+                            case 'text':
+                                nestedQuery = {
+                                    '_rule_type': 'containment',
+                                    'contains': [
+                                        value
+                                    ]
+                                };
+                                break;
+
+                            case 'select':
```

Comment:
Agreed, although it seems to me like perhaps those two design processes (JS library + JSON query API) could happen coterminously? It seems to me at least like a primary goal of the JS library would be to put an abstraction layer over the query API so that the developer doesn't have to think too hard about how querying works -- although it's true that lessons learned from improving the query API would likely trickle down to the JS API, too. In any case, it's fun to think about, and I'm looking forward to https://github.com/azavea/ashlar-2018-fellowship/issues/20 and https://github.com/azavea/ashlar-2018-fellowship/issues/16.

Reply:
Yup, totally agree, I think it can go both ways.